### ISSUE TITLE
[논문 플랜] Consistency × RL × Meta-Cognition × Reasoning 통합 — 실험 구성 및 탑컨퍼 비판적 평가

### ISSUE THREAD

**Issue by kimshhhhh** [kimshhhhh/research#7]
# Consistency × RL × Meta-Cognition × Reasoning 통합 논문 플랜

**Research Date**: 2026-04-17  
**Related Issues**: [#5 (논문 서베이)](https://github.com/kimshhhhh/research/issues/5)  
**Target Conference**: NeurIPS 2026 / ICLR 2026  
**Status**: 🔴 Draft — 비판적 검토 진행 중

---

## 🎯 핵심 프레이밍: 왜 이 주제인가

NeurIPS 2025 Best Paper Runner-Up ([arXiv:2504.13837](https://arxiv.org/abs/2504.13837))의 핵심 발견:

> "RLVR은 base 모델에 이미 존재하는 reasoning path를 재가중할 뿐, **reasoning capability boundary를 확장하지 못한다**"

이 발견이 열어둔 **open question**:
- RL이 boundary를 못 넓힌다면, 모델이 자신의 boundary를 **정확히 인식**하도록 만들 수는 없을까?
- Boundary를 인식하면 → 적절한 fallback (검색, 도구, human escalation) 가능
- 이것이 hallucination 감소, agent self-routing의 근본 해법

**핵심 프레이밍 전환**:  
`boundary expansion (불가능)` → `boundary awareness (우리의 목표)`

---

## 📌 제안 제목

> **"Know Your Limits: Meta-Cognitive Reinforcement Learning with Consistency-Based Boundary Detection for LLM Reasoning"**

단축 제목 (제출용): **MC-BDR** (Meta-Cognitive Boundary-aware reasoning via RL)

---

### Stage 1 — Boundary Recognition (메타인지 예측)

문제를 받으면 reasoning 전에 먼저 solvability를 예측:

```
Input problem p
→ Meta-cognitive head: P(solvable | p, model_state) ∈ [0, 1]
```

- Softmax 기반 confidence가 아닌 **reasoning trace를 생성한 후** self-evaluation
- 핵심: "이 문제가 내 분포 안에 있는가"를 reasoning 과정 자체로 판단
- 구현: 경량 linear head on last-token hidden state (추론 시 overhead 최소화)

### Stage 2 — Consistency-Based Verification (경계 탐지)

Stage 1에서 `P(solvable)` 이 [0.3, 0.7] 범위 (uncertain zone)인 문제에만 적용:

```
Uncertain problem → Sample N reasoning paths {τ_1, ..., τ_N}
→ Consistency score: C = agreement_rate(final_answers)
→ High C (>0.8): boundary 내부 → re-classify as solvable
→ Low C (<0.3): boundary 외부 신호 → flag as OOB
```

- **Selective compute**: 모든 문제가 아닌 uncertain zone에만 N-path sampling 적용
- 이 자체가 contribution: adaptive consistency verification (vs. 모든 문제에 self-consistency 적용하는 기존 방법)

### Stage 3 — RL with Meta-Cognitive Reward

기존 RLVR reward: `r = 1 if correct else 0`

**제안 reward**:
```python
r_total = r_answer + α × r_calibration + β × r_consistency_signal

# r_calibration: 메타인지 정확도 보상
r_calibration = -|P_predicted(solvable) - I(actually_correct)|
# → ECE (Expected Calibration Error) 최소화 방향

# r_consistency_signal: boundary 외부 문제를 정확히 flagging했는지
r_consistency_signal = +1 if (C_low AND actually_wrong)   # TP
                     = -1 if (C_low AND actually_correct)  # FP (false alarm)
                     = 0  otherwise
```

하이퍼파라미터: `α, β`는 curriculum에 따라 점진적 증가

---

### 핵심 실험 1: Boundary Detection Accuracy

**목표**: 모델이 풀 수 있는/없는 문제를 얼마나 정확히 구분하는가

**Ground truth 정의**:
- 동일 문제를 base 모델로 k=32번 샘플링
- `pass@32 > 0.5` → "solvable" (boundary 내부)
- `pass@32 < 0.1` → "unsolvable" (boundary 외부)
- 나머지 → "boundary region" (uncertain)

**측정 지표**:
- AUC-ROC (solvable vs. unsolvable 분류)
- ECE (Expected Calibration Error)
- Boundary region accuracy (가장 어려운 케이스)

**비교 baseline**:
- Raw confidence (softmax probability)
- Self-consistency (기존 방식, 모든 문제에 N=10 path)
- MASA-RL의 meta-awareness signal

### 핵심 실험 2: Consistency ↔ Boundary 상관관계 검증

**목표**: "consistency가 낮은 문제 = boundary 근처 문제"라는 가설 검증

**방법**:
1. 문제를 boundary 위치 기준으로 3구간 분류 (inner / boundary / outer)
2. 각 구간에서 N-path consistency 분포 측정
3. Spearman correlation 계산

**기대 결과**: boundary 구간에서 consistency 분산 최대, inner에서 최소  
**만약 상관관계가 낮다면**: 이 자체도 중요한 발견 (negative result도 출판 가능)

### 핵심 실험 3: MC-BDR vs. Existing RL Methods

**목표**: 제안 방법이 기존 RL 대비 calibration + accuracy 모두 개선하는지 검증

| 비교 대상 | 논문 |
|-----------|------|
| RLVR (verifiable reward) | Does RL Really... 기준 |
| SCoRe (multi-turn self-correction) | ICLR 2025 Oral |
| CoVo (consistency as reward) | NeurIPS 2025 |
| MASA-RL (meta-awareness RL) | arXiv 2025 |
| **MC-BDR (ours)** | — |

**측정 지표**:
- Accuracy (primary)
- ECE (calibration quality)
- Abstention quality: "모른다"고 할 때 실제로 틀린 비율
- Compute efficiency: uncertain zone detection이 전체 compute를 얼마나 절감하는가

### 핵심 실험 4: OOD Generalization

**목표**: boundary awareness가 분포 외 문제에서도 작동하는가 (진짜 novelty)

- Train: MATH
- Test: AIME 2025, GPQA-Diamond, LiveCodeBench
- **핵심 가설**: boundary awareness를 훈련한 모델은 OOD에서도 calibration이 유지됨
- 이것이 "boundary expansion 불가" 논문에 대한 우리의 empirical 반박

### 핵심 실험 5: Ablation Study

| 제거 구성요소 | 기대 저하 |
|-------------|---------|
| r_calibration 제거 | ECE 악화 |
| r_consistency_signal 제거 | boundary detection AUC 저하 |
| Selective consistency (uncertain zone only) → Full consistency | compute 증가, accuracy 유사 |
| Stage 2 (consistency verification) 제거 | boundary region accuracy 저하 |

---

## 📊 벤치마크 선택 및 근거

| 벤치마크 | 역할 | 근거 |
|---------|------|------|
| **MATH** (Hendrycks 2021) | 주 훈련/평가 | boundary 분포가 풍부, 기존 연구 표준 |
| **AIME 2024/2025** | 고난도 OOD | boundary 외부 문제 비율 높음, NeurIPS 논문 사용 |
| **GPQA-Diamond** | 과학 분야 | 전문가도 어려운 경계 문제 풍부 |
| **LiveCodeBench** | 시간적 OOD | 훈련 후 등장한 문제 → boundary shift 관찰 |
| **GSM8K** | 쉬운 분포 | boundary 내부 문제가 많음 → calibration 안정성 확인 |

---

### Phase 0: 환경 구축 (1주)
- Base model: Qwen2.5-7B-Instruct (compute 효율, 오픈소스)
- RL framework: verl 또는 trl (GRPO 지원)
- 벤치마크 데이터 전처리 및 boundary ground truth 생성

### Phase 1: Boundary Ground Truth 생성 (2주)
- MATH train set에 대해 pass@32 계산 (base model)
- 3구간 분류 레이블 생성
- Consistency ↔ boundary 상관관계 예비 분석 (실험 2 선행)

### Phase 2: Stage 1+2 구현 (2주)
- Meta-cognitive head 구현 및 훈련
- Selective consistency verification 구현
- Boundary detection accuracy baseline 측정

### Phase 3: RL 훈련 (3주)
- Reward 함수 구현 (r_answer + α×r_calibration + β×r_consistency_signal)
- Curriculum 설계: α, β 점진적 증가
- 훈련 안정성 모니터링

### Phase 4: 전체 실험 + 분석 (2주)
- 실험 1~4 완전 실행
- Ablation study
- Baseline 비교 실험

### Phase 5: 논문 작성 (2주)
- NeurIPS 2026 데드라인: ~2026년 5월 (예상)

**총 예상 기간**: 약 12주 (2026-04-17 ~ 2026-07-10)

---

### 🔴 치명적 리스크

**1. MASA-RL과의 차별화 문제**  
[MASA-RL (arXiv:2510.03259)](https://arxiv.org/abs/2510.03259)도 meta-awareness + RL을 결합함.  
차이: MASA-RL은 meta-awareness와 reasoning rollout의 **정렬**에 집중, consistency 미통합.  
→ **대응**: "meta-awareness의 calibration target으로 consistency를 사용한다"는 점을 명확히 framing. MASA-RL과의 직접 비교 실험 필수.

**2. "그냥 calibration 논문" 반박 위험**  
리뷰어가 "이건 그냥 ECE 최소화 논문 아닌가?"라고 할 수 있음.  
→ **대응**: Calibration은 post-hoc 보정. 우리는 RL로 **reasoning 과정에서 동적으로** boundary를 인식하도록 훈련. 이론적으로 명확히 구분 필요.

**3. Compute 오버헤드 정당화**  
Uncertain zone에만 N-path sampling을 해도, 추론 시 Stage 1+2 모두 실행해야 함.  
→ **대응**: 전체 throughput이 표준 self-consistency (모든 문제에 N=10) 대비 낮아야 함. 실측 필수.

### 🟡 중간 리스크

**4. RL 훈련 불안정성**  
3개 reward term (r_answer + r_calibration + r_consistency_signal) 동시 최적화.  
→ **대응**: 단계적 curriculum (먼저 r_answer만, 이후 calibration term 추가). Group Relative Policy Optimization (GRPO) 사용 시 상대적으로 안정적.

**5. Pass@32 ground truth의 circular argument 문제**  
Base 모델로 boundary ground truth를 만들고, 그 base 모델을 RL로 훈련.  
→ "RL 이후 boundary가 달라지면?" 이슈 존재.  
→ **대응**: Evaluation 시에는 **새로운 모델 checkpoint**로 boundary를 재정의하지 않음. Ground truth는 고정된 base 모델 기준으로만.

**6. Boundary ground truth의 연속성 무시**  
Pass@32로 이진 분류하지만 실제 boundary는 연속적.  
→ **대응**: 민감도 분석 (threshold 변화에 따른 결과 안정성 확인)

### 🟢 핸들 가능한 리스크

**7. "왜 consistency가 boundary indicator인가" 이론 부재**  
→ Informal justification으로도 충분. NeurIPS는 empirical 결과가 강하면 통과 가능.  
단, information-theoretic 관점의 짧은 theoretical section 추가 권장.

---

### 강점 (탑컨퍼 어필 포인트)

| 어필 포인트 | 구체적 근거 |
|------------|-----------|
| **Timely framing** | NeurIPS 2025 Best Paper Runner-Up 한계를 직접 address |
| **명확한 연구 공백** | Issue #5 서베이에서 확인: 4개 교차점 통합 연구 없음 |
| **실용적 임팩트** | Hallucination 감소, agent self-routing — AI safety 연결 가능 |
| **Negative result 활용** | "RL은 boundary를 못 넓힌다" → "대신 인식하게 만든다" (elegant reframing) |

### 예상 억셉 확률

| 컨퍼 | 확률 | 조건 |
|------|------|------|
| **NeurIPS 2026** | ~30-35% | 실험 4개 모두 강한 결과, MASA-RL 비교 필수 |
| **ICLR 2026** | ~25-30% | Calibration + reasoning 교차점 well-received |
| **ICML 2026** | ~20-25% | RL theory 섹션 추가 시 유리 |

> **솔직한 평가**: 아이디어 자체는 탑컨퍼급. 하지만 MASA-RL과의 차별화, compute 오버헤드 실측, calibration 논문과의 구분 이 3가지가 리뷰에서 집중 공격받을 것. 이 3가지를 철저히 준비하면 NeurIPS 2026 **현실적으로 도전 가능**.

---

## 📋 즉시 해야 할 일 (To-Do)

- [ ] MASA-RL 코드([akatigre/MASA-RL](https://github.com/akatigre/MASA-RL)) 분석 — 차별화 포인트 구체화
- [ ] CoVo 코드([sastpg/CoVo](https://github.com/sastpg/CoVo)) 분석 — consistency reward 구현 방식 참조
- [ ] Base model 선택 확정 (Qwen2.5-7B vs. LLaMA3.1-8B)
- [ ] MATH train set pass@32 계산 시작 (Phase 1)
- [ ] kimshhhhh/consistency-reasoning 레포에 구현 이슈 작성

---

*Related Issues: [research#5 (논문 서베이)](https://github.com/kimshhhhh/research/issues/5) | [research#3 (주제 선정)](https://github.com/kimshhhhh/research/issues/3)*  
*Research plan drafted by: oh-my-agents | 2026-04-17*
